### DISCUSSION TITLE
My training is failing despite my images looking fine

### DISCUSSION THREAD

**Discussion by sp01186** [facebookresearch/detectron2#4270] [Q&A]
My training isn't doing what I want it to do and I don't understand why, I was hoping I could get some explanation or support with this from this forum. 

![image](https://user-images.githubusercontent.com/58943663/170110363-4816825e-9ffc-4ded-8d15-fe372811bbaa.png)

As you can see from the image above, the data seems to be okay in its identification of the bounding boxes.

The training code is shown here. 

![image](https://user-images.githubusercontent.com/58943663/170110775-9a2dadbb-e090-4caf-b603-a7520084640c.png)

 However after I train them, the predictions I am getting are all over the place. 

![image](https://user-images.githubusercontent.com/58943663/170111057-4cfa5e23-6986-4fc0-bf57-e6bb87b6ca12.png)

![image](https://user-images.githubusercontent.com/58943663/170111164-2ffeef48-4d45-4da7-b079-1891ed2e8743.png)

Please let me know if there are any suggestions or solutions.

**Comment by ExtReMLapin** [comment]
Can you share your training json file ? and the images ?

**Reply by sp01186** [reply]
It turns out that the problem was that I didn't have enough training iterations. I thought that it would at least give something that wasn't completely random if I used 300 iterations,. 

The training data can be obtained with the following: 

  !pip install "opencv-python-headless<4.3"
  !pip install --upgrade pip setuptools wheel
  !pip install fiftyone
  
  import fiftyone as fo
  import fiftyone.zoo as foz
  import numpy as np
  import tensorflow as tf

# Download and load the validation split of Open Images V6
#dataset = foz.load_zoo_dataset("open-images-v6", split="validation") Full dataset.
dataset = foz.load_zoo_dataset(
              "open-images-v6",
              split="train",
              label_types=["detections", "classifications", "segmentations"],
              classes=["Airplane","Cat", "Dog"],
              max_samples=20000
          )

# Converting the dataset into the detectron2 format

from detectron2.structures import BoxMode

def get_data_dicts(dataset):

    dataset.compute_metadata()
    dataset_dicts = []

    for sample in dataset.iter_samples():
    # for idx, v in enumerate(imgs_anns.values()):
        record = {}

        filename = sample.filepath
        idx = sample.id
        
        record["file_name"] = filename # found
        record["image_id"] = idx # found
        

        detections = sample.detections;

        classes = ['Airplane', 'Dog', 'Cat'];

        
        objs = []
        for i in detections.detections:
        # print(i)
        # print(i.label)
          if i.label in classes:
            coors = []
            category = 0
            if i.label == "Dog":
              category = 1
            if i.label == "Cat":
              category = 2
            #print(i.bounding_box) #will need to convert from relative to pixel coordinates
            for indice in range(4):
              if (indice == 0 or indice == 2):
                coors.append(i.bounding_box[indice]*sample.metadata.width)
              else:
                coors.append(i.bounding_box[indice]*sample.metadata.height)
            obj = {
                "bbox": coors, # found
                "bbox_mode": BoxMode.XYWH_ABS, #default
                # "segmentation": [poly],
                "category_id": category,
              }
            objs.append(obj)
        record["annotations"] = objs
        dataset_dicts.append(record)

    return dataset_dicts
